pypeit/pypeit.py

```diff
@@ -566,84 +575,59 @@ def extract_one(self, frames, det, bg_frames=[], std_outfile=None):
-
-            self.skymodel, self.objmodel, self.ivarmodel, self.outmask, self.sobjs = \
-            self.redux.local_skysub_extract(self.caliBrate.mswave, self.global_sky, self.sobjs_obj,
-                                            model_noise=(not self.ir_redux),std = self.std_redux,
-                                            maskslits=self.maskslits, show_profile=self.show,show=self.show)
-
-            # Purge out the negative objects if this was a near-IR reduction.
-            # TODO should we move this purge call to local_skysub_extract??
-            if self.ir_redux:
-                self.sobjs.purge_neg()
+        self.skymodel, self.objmodel, self.ivarmodel, self.outmask, self.sobjs = self.redux.run(
+            std_trace=std_trace, manual_extract_dict=manual_extract_dict,
+            show_peaks=self.show)
 
-            # Flexure correction if this is not a standard star
-            if not self.std_redux:
-                self.redux.flexure_correct(self.sobjs, self.basename)
+        '''
+        # Find objects
+        self.sobjs_obj, self.nobj, self.skymask = self.redux.find_objects(
+            std_trace=std_trace, manual_extract_dict=manual_extract_dict)
 
-            # Grab coord
-            radec = ltu.radec_to_coord((self.fitstbl["ra"][frames[0]], self.fitstbl["dec"][frames[0]]))
-            self.redux.helio_correct(self.sobjs, radec, self.obstime)
-            #embed(header='620 of pypeit')
+        # Extract (if nobj==0, this passes back 'dummy' arrays required for spec2d outputs)
+        self.skymodel, self.objmodel, self.ivarmodel, self.outmask, self.sobjs = self.redux.extract()
+        '''
 
-        else:
+        # Finish up
+        if self.sobjs.nobj == 0:
             # Print status message
-            msgs_string = 'No objects to extract for target {:s}'.format(self.fitstbl['target'][frames[0]]) + msgs.newline()
+            msgs_string = 'No objects to extract for file {:s}'.format(
+                self.fitstbl['target'][frames[0]]) + msgs.newline()
             msgs_string += 'On frames:' + msgs.newline()
             for iframe in frames:
                 msgs_string += '{0:s}'.format(self.fitstbl['filename'][iframe]) + msgs.newline()
             msgs.warn(msgs_string)
-            # set to first pass global sky
-            self.skymodel = self.initial_sky
-            self.objmodel = np.zeros_like(self.sciImg.image)
-            # Set to sciivar. Could create a model but what is the point?
-            self.ivarmodel = np.copy(self.sciImg.ivar)
-            # Set to the initial mask in case no objects were found
-            self.outmask = self.sciImg.mask
-            # empty specobjs object from object finding
-            if self.ir_redux:
-                self.sobjs_obj.purge_neg()
-            self.sobjs = self.sobjs_obj
+        else:
+            # TODO -- Should we move these to redux.run()?
```

Comment:
It seems preferable to do execute those steps there, however the annoying thing is that you need basename,radec, and obstime. I'm sure your solution would be to add fitstable as an argument to reduce, which I would violently oppose. 

We can discuss this. In some cases you may want to reduce without performing flexure and helio, so perhaps these should optional argument, or perhaps we just leave them here.

Reply:
moved